File: app/routes/drivers.py
from flask import Blueprint, request, jsonify, render_template, flash, redirect, url_for
from flask_login import login_required, current_user
from app.decorators import operations_required
from app.models.samsara import SamsaraDriver, SamsaraAlert, SamsaraVehicle
from app.models.company import Company
from app import db
import logging
from datetime import datetime, timedelta
from sqlalchemy import or_, func

logger = logging.getLogger(__name__)
bp = Blueprint('drivers', __name__, url_prefix='/drivers')

@bp.route('/')
@login_required
@operations_required
def index():
    """Driver management page"""
    return render_template('drivers/index.html')

@bp.route('/api/drivers')
@login_required
@operations_required
def get_drivers():
    """Get list of drivers with filters and pagination"""
    try:
        page = request.args.get('page', 1, type=int)
        per_page = request.args.get('per_page', 20, type=int)
        search = request.args.get('search')
        company_id = request.args.get('company_id', type=int)
        status = request.args.get('status')  # active, inactive
        sort_by = request.args.get('sort_by', 'name')
        sort_order = request.args.get('sort_order', 'asc')

        # Start with base query
        query = SamsaraDriver.query.options(
            db.joinedload(SamsaraDriver.company)
        )

        # Apply filters
        if search:
            search_term = f"%{search}%"
            query = query.filter(or_(
                SamsaraDriver.name.ilike(search_term),
                SamsaraDriver.username.ilike(search_term),
                SamsaraDriver.phone.ilike(search_term),
                SamsaraDriver.email.ilike(search_term)
            ))

        if company_id:
            query = query.filter(SamsaraDriver.company_id == company_id)

        if status:
            if status == 'active':
                query = query.filter(SamsaraDriver.is_active == True)
            elif status == 'inactive':
                query = query.filter(SamsaraDriver.is_active == False)

        # Apply sorting
        valid_sort_fields = {
            'name': SamsaraDriver.name,
            'created_at': SamsaraDriver.created_at,
            'updated_at': SamsaraDriver.updated_at,
            'company_name': Company.name
        }
        
        if sort_by in valid_sort_fields:
            sort_field = valid_sort_fields[sort_by]
            if sort_by == 'company_name':
                query = query.join(Company)
                
            if sort_order.lower() == 'desc':
                query = query.order_by(sort_field.desc())
            else:
                query = query.order_by(sort_field.asc())
        else:
            query = query.order_by(SamsaraDriver.name.asc())

        # Paginate results
        pagination = query.paginate(page=page, per_page=per_page, error_out=False)
        
        # Format driver data
        drivers_data = []
        for driver in pagination.items:
            try:
                # Get alert statistics safely
                total_alerts = SamsaraAlert.query.filter_by(driver_id=driver.id).count()
                recent_alerts = SamsaraAlert.query.filter(
                    SamsaraAlert.driver_id == driver.id,
                    SamsaraAlert.created_at >= datetime.utcnow() - timedelta(days=30)
                ).count()
                
                drivers_data.append({
                    'id': driver.id,
                    'driver_id': driver.driver_id,
                    'name': driver.name,
                    'username': driver.username,
                    'phone': driver.phone,
                    'email': driver.email,
                    'license_number': driver.license_number,
                    'license_state': driver.license_state,
                    'license_class': driver.license_class,
                    'company_name': driver.company.name if driver.company else 'No Company',
                    'company_id': driver.company_id,
                    'is_active': driver.is_active,
                    'created_at': driver.created_at.isoformat() if driver.created_at else None,
                    'updated_at': driver.updated_at.isoformat() if driver.updated_at else None,
                    'total_alerts': total_alerts,
                    'recent_alerts': recent_alerts
                })
            except Exception as driver_error:
                logger.error(f"Error processing driver {driver.id}: {str(driver_error)}")
                # Add driver with minimal data if there's an error
                drivers_data.append({
                    'id': driver.id,
                    'driver_id': driver.driver_id,
                    'name': driver.name or 'Unknown',
                    'username': driver.username,
                    'phone': driver.phone,
                    'email': driver.email,
                    'license_number': driver.license_number,
                    'license_state': driver.license_state,
                    'license_class': driver.license_class,
                    'company_name': driver.company.name if driver.company else 'No Company',
                    'company_id': driver.company_id,
                    'is_active': driver.is_active,
                    'created_at': driver.created_at.isoformat() if driver.created_at else None,
                    'updated_at': driver.updated_at.isoformat() if driver.updated_at else None,
                    'total_alerts': 0,
                    'recent_alerts': 0
                })
        
        return jsonify({
            'status': 'success',
            'drivers': drivers_data,
            'total': pagination.total,
            'pages': pagination.pages,
            'current_page': page,
            'per_page': per_page
        })
        
    except Exception as e:
        logger.error(f"Error fetching drivers: {str(e)}", exc_info=True)
        return jsonify({
            'status': 'error',
            'message': f'Failed to fetch drivers: {str(e)}'
        }), 500

@bp.route('/api/drivers/<int:driver_id>')
@login_required
@operations_required
def get_driver_details(driver_id):
    """Get detailed information for a specific driver"""
    try:
        driver = SamsaraDriver.query.options(
            db.joinedload(SamsaraDriver.company)
        ).get_or_404(driver_id)
        
        # Get alert statistics safely
        alerts_query = SamsaraAlert.query.filter_by(driver_id=driver.id)
        total_alerts = alerts_query.count()
        
        # Alert counts by status
        alert_stats = {
            'total': total_alerts,
            'unassigned': alerts_query.filter(SamsaraAlert.status == 'unassigned').count(),
            'in_progress': alerts_query.filter(SamsaraAlert.status == 'in_progress').count(),
            'resolved': alerts_query.filter(SamsaraAlert.status == 'resolved').count(),
            'escalated': alerts_query.filter(SamsaraAlert.status == 'escalated').count()
        }
        
        # Recent alerts (last 10)
        recent_alerts = alerts_query.order_by(SamsaraAlert.created_at.desc()).limit(10).all()
        recent_alerts_data = []
        for alert in recent_alerts:
            try:
                recent_alerts_data.append({
                    'id': alert.id,
                    'alert_id': alert.alert_id,
                    'alert_type': alert.alert_type,
                    'severity': alert.severity,
                    'status': alert.status,
                    'vehicle_name': alert.vehicle.name if alert.vehicle else 'Unknown',
                    'timestamp': alert.timestamp.isoformat() if alert.timestamp else None,
                    'created_at': alert.created_at.isoformat() if alert.created_at else None
                })
            except Exception as alert_error:
                logger.error(f"Error processing alert {alert.id}: {str(alert_error)}")
                # Add alert with minimal data
                recent_alerts_data.append({
                    'id': alert.id,
                    'alert_id': alert.alert_id,
                    'alert_type': alert.alert_type,
                    'severity': alert.severity,
                    'status': alert.status,
                    'vehicle_name': 'Unknown',
                    'timestamp': alert.timestamp.isoformat() if alert.timestamp else None,
                    'created_at': alert.created_at.isoformat() if alert.created_at else None
                })
        
        driver_data = {
            'id': driver.id,
            'driver_id': driver.driver_id,
            'name': driver.name,
            'username': driver.username,
            'phone': driver.phone,
            'email': driver.email,
            'license_number': driver.license_number,
            'license_state': driver.license_state,
            'license_class': driver.license_class,
            'company': {
                'id': driver.company.id,
                'name': driver.company.name
            } if driver.company else None,
            'external_ids': driver.external_ids,
            'is_active': driver.is_active,
            'created_at': driver.created_at.isoformat() if driver.created_at else None,
            'updated_at': driver.updated_at.isoformat() if driver.updated_at else None,
            'alert_stats': alert_stats,
            'recent_alerts': recent_alerts_data,
            'samsara_data': driver.data  # Full Samsara data
        }
        
        return jsonify({
            'status': 'success',
            'driver': driver_data
        })
        
    except Exception as e:
        logger.error(f"Error fetching driver details: {str(e)}", exc_info=True)
        return jsonify({
            'status': 'error',
            'message': f'Failed to fetch driver details: {str(e)}'
        }), 500

@bp.route('/api/companies')
@login_required
@operations_required
def get_companies():
    """Get list of companies for filter dropdown"""
    try:
        # Get companies that have drivers
        companies = Company.query.filter(
            Company.id.in_(
                db.session.query(SamsaraDriver.company_id).filter(
                    SamsaraDriver.company_id.isnot(None)
                ).distinct()
            )
        ).all()
        
        companies_data = []
        for company in companies:
            try:
                driver_count = SamsaraDriver.query.filter_by(
                    company_id=company.id, 
                    is_active=True
                ).count()
                companies_data.append({
                    'id': company.id,
                    'name': company.name,
                    'driver_count': driver_count
                })
            except Exception as company_error:
                logger.error(f"Error processing company {company.id}: {str(company_error)}")
                # Add company with minimal data
                companies_data.append({
                    'id': company.id,
                    'name': company.name,
                    'driver_count': 0
                })
        
        return jsonify({
            'status': 'success',
            'companies': companies_data
        })
    except Exception as e:
        logger.error(f"Error fetching companies: {str(e)}", exc_info=True)
        return jsonify({
            'status': 'error',
            'message': f'Failed to fetch companies: {str(e)}'
        }), 500

@bp.route('/<int:driver_id>')
@login_required
@operations_required
def driver_details(driver_id):
    """Driver details page"""
    try:
        driver = SamsaraDriver.query.get_or_404(driver_id)
        return render_template('drivers/details.html', driver=driver)
    except Exception as e:
        logger.error(f"Error loading driver details page: {str(e)}")
        flash('Driver not found', 'error')
        return redirect(url_for('drivers.index')) 

File: app/static/js/drivers.js
// Driver Management JavaScript

let currentPage = 1;
let currentFilters = {};
let currentDriverId = null;

// Initialize page
document.addEventListener('DOMContentLoaded', function() {
    loadCompanies();
    loadDrivers();
    setupEventListeners();
});

// Setup event listeners
function setupEventListeners() {
    // Search input
    const searchInput = document.getElementById('searchDrivers');
    let searchTimeout;
    searchInput.addEventListener('input', function() {
        clearTimeout(searchTimeout);
        searchTimeout = setTimeout(() => {
            currentPage = 1;
            loadDrivers();
        }, 500);
    });

    // Filter dropdowns
    ['companyFilter', 'statusFilter', 'sortBy', 'sortOrder'].forEach(id => {
        document.getElementById(id).addEventListener('change', function() {
            currentPage = 1;
            loadDrivers();
        });
    });
}

// Load companies for filter dropdown
async function loadCompanies() {
    try {
        const response = await fetch('/drivers/api/companies');
        const data = await response.json();
        
        if (data.status === 'success') {
            const companySelect = document.getElementById('companyFilter');
            companySelect.innerHTML = '<option value="">All Companies</option>';
            
            data.companies.forEach(company => {
                const option = document.createElement('option');
                option.value = company.id;
                option.textContent = `${company.name} (${company.driver_count} drivers)`;
                companySelect.appendChild(option);
            });
        }
    } catch (error) {
        console.error('Error loading companies:', error);
    }
}

// Load drivers with current filters
async function loadDrivers() {
    showLoading(true);
    
    try {
        // Build query parameters
        const params = new URLSearchParams({
            page: currentPage,
            per_page: 20,
            search: document.getElementById('searchDrivers').value,
            company_id: document.getElementById('companyFilter').value,
            status: document.getElementById('statusFilter').value,
            sort_by: document.getElementById('sortBy').value,
            sort_order: document.getElementById('sortOrder').value
        });

        // Remove empty parameters
        for (let [key, value] of [...params.entries()]) {
            if (!value) params.delete(key);
        }

        const response = await fetch(`/drivers/api/drivers?${params}`);
        const data = await response.json();
        
        if (data.status === 'success') {
            renderDrivers(data.drivers);
            renderPagination(data.current_page, data.pages, data.total);
            updateDriversCount(data.total);
        } else {
            showError('Failed to load drivers');
        }
    } catch (error) {
        console.error('Error loading drivers:', error);
        showError('Failed to load drivers');
    } finally {
        showLoading(false);
    }
}

// Render drivers table
function renderDrivers(drivers) {
    const tbody = document.getElementById('driversTableBody');
    const emptyState = document.getElementById('driversEmpty');
    
    if (!drivers || drivers.length === 0) {
        tbody.innerHTML = '';
        emptyState.style.display = 'block';
        return;
    }
    
    emptyState.style.display = 'none';
    
    tbody.innerHTML = drivers.map(driver => {
        const avatarColor = getAvatarColor(driver.name);
        const statusClass = driver.is_active ? 'driver-status-active' : 'driver-status-inactive';
        const statusIcon = driver.is_active ? 'fa-check-circle' : 'fa-times-circle';
        
        return `
            <tr class="driver-row" onclick="showDriverDetails(${driver.id})">
                <td>
                    <div class="d-flex align-items-center">
                        <div class="driver-avatar" style="background-color: ${avatarColor};">
                            ${(driver.name || 'D')[0].toUpperCase()}
                        </div>
                        <div>
                            <div class="fw-bold">${driver.name || 'Unknown Driver'}</div>
                            <small class="text-muted">
                                ID: ${driver.driver_id}
                                ${driver.username ? ` • @${driver.username}` : ''}
                            </small>
                        </div>
                    </div>
                </td>
                <td>
                    <i class="fas ${statusIcon} ${statusClass} me-1"></i>
                    <span class="${statusClass}">${driver.is_active ? 'Active' : 'Inactive'}</span>
                </td>
                <td>
                    <div>
                        ${driver.phone ? `<div><i class="fas fa-phone me-1"></i>${driver.phone}</div>` : ''}
                        ${driver.email ? `<div class="text-muted small"><i class="fas fa-envelope me-1"></i>${driver.email}</div>` : ''}
                        ${!driver.phone && !driver.email ? '<span class="text-muted">No contact info</span>' : ''}
                    </div>
                </td>
                <td>
                    <div>
                        ${driver.license_number ? `<div>${driver.license_number}</div>` : ''}
                        ${driver.license_state ? `<small class="text-muted">${driver.license_state}${driver.license_class ? ` • ${driver.license_class}` : ''}</small>` : ''}
                        ${!driver.license_number && !driver.license_state ? '<span class="text-muted">N/A</span>' : ''}
                    </div>
                </td>
                <td>
                    <span class="badge bg-light text-dark">
                        <i class="fas fa-building me-1"></i>
                        ${driver.company_name}
                    </span>
                </td>
                <td>
                    <div class="alert-stats">
                        <span class="badge bg-primary">${driver.total_alerts}</span>
                        ${driver.recent_alerts > 0 ? `<span class="badge bg-warning">${driver.recent_alerts} recent</span>` : ''}
                    </div>
                </td>
                <td>
                    <small class="text-muted">
                        ${formatDate(driver.created_at)}
                    </small>
                </td>
                <td>
                    <div class="btn-group">
                        <button class="btn btn-sm btn-outline-primary" onclick="event.stopPropagation(); showDriverDetails(${driver.id})">
                            <i class="fas fa-eye"></i>
                        </button>
                        <button class="btn btn-sm btn-outline-secondary" onclick="event.stopPropagation(); viewDriverPage(${driver.id})">
                            <i class="fas fa-external-link-alt"></i>
                        </button>
                    </div>
                </td>
            </tr>
        `;
    }).join('');
}

// Show driver details in modal
async function showDriverDetails(driverId) {
    currentDriverId = driverId;
    
    try {
        const response = await fetch(`/drivers/api/drivers/${driverId}`);
        const data = await response.json();
        
        if (data.status === 'success') {
            renderDriverDetailsModal(data.driver);
            const modal = new bootstrap.Modal(document.getElementById('driverDetailsModal'));
            modal.show();
        } else {
            showError('Failed to load driver details');
        }
    } catch (error) {
        console.error('Error loading driver details:', error);
        showError('Failed to load driver details');
    }
}

// Render driver details in modal
function renderDriverDetailsModal(driver) {
    const container = document.getElementById('driverDetailsContent');
    const avatarColor = getAvatarColor(driver.name);
    
    container.innerHTML = `
        <div class="row">
            <!-- Driver Header -->
            <div class="col-12 mb-4">
                <div class="d-flex align-items-center p-3 bg-light rounded">
                    <div class="driver-avatar me-3" style="background-color: ${avatarColor}; width: 60px; height: 60px; font-size: 1.5rem;">
                        ${(driver.name || 'D')[0].toUpperCase()}
                    </div>
                    <div class="flex-grow-1">
                        <h4 class="mb-1">${driver.name || 'Unknown Driver'}</h4>
                        <p class="mb-1 text-muted">
                            <i class="fas fa-id-card me-2"></i>Driver ID: ${driver.driver_id}
                            ${driver.username ? ` • @${driver.username}` : ''}
                        </p>
                        <div class="d-flex gap-2">
                            <span class="badge ${driver.is_active ? 'bg-success' : 'bg-danger'}">
                                ${driver.is_active ? 'Active' : 'Inactive'}
                            </span>
                            ${driver.company ? `<span class="badge bg-light text-dark">${driver.company.name}</span>` : ''}
                        </div>
                    </div>
                </div>
            </div>
            
            <!-- Statistics -->
            <div class="col-12 mb-4">
                <h6>Alert Statistics</h6>
                <div class="row g-2">
                    <div class="col-3">
                        <div class="text-center p-2 bg-primary text-white rounded">
                            <div class="h4 mb-0">${driver.alert_stats.total}</div>
                            <small>Total</small>
                        </div>
                    </div>
                    <div class="col-3">
                        <div class="text-center p-2 bg-warning text-white rounded">
                            <div class="h4 mb-0">${driver.alert_stats.unassigned}</div>
                            <small>Unassigned</small>
                        </div>
                    </div>
                    <div class="col-3">
                        <div class="text-center p-2 bg-info text-white rounded">
                            <div class="h4 mb-0">${driver.alert_stats.in_progress}</div>
                            <small>In Progress</small>
                        </div>
                    </div>
                    <div class="col-3">
                        <div class="text-center p-2 bg-success text-white rounded">
                            <div class="h4 mb-0">${driver.alert_stats.resolved}</div>
                            <small>Resolved</small>
                        </div>
                    </div>
                </div>
            </div>
            
            <!-- Driver Information -->
            <div class="col-md-6">
                <h6>Contact Information</h6>
                <table class="table table-sm">
                    <tr>
                        <td><strong>Phone:</strong></td>
                        <td>${driver.phone || 'N/A'}</td>
                    </tr>
                    <tr>
                        <td><strong>Email:</strong></td>
                        <td>${driver.email || 'N/A'}</td>
                    </tr>
                </table>
                
                <h6 class="mt-3">License Information</h6>
                <table class="table table-sm">
                    <tr>
                        <td><strong>License Number:</strong></td>
                        <td>${driver.license_number || 'N/A'}</td>
                    </tr>
                    <tr>
                        <td><strong>State:</strong></td>
                        <td>${driver.license_state || 'N/A'}</td>
                    </tr>
                    <tr>
                        <td><strong>Class:</strong></td>
                        <td>${driver.license_class || 'N/A'}</td>
                    </tr>
                </table>
            </div>
            
            <!-- Recent Alerts -->
            <div class="col-md-6">
                <h6>Recent Alerts</h6>
                <div style="max-height: 300px; overflow-y: auto;">
                    ${driver.recent_alerts && driver.recent_alerts.length > 0 ? 
                        driver.recent_alerts.map(alert => `
                            <div class="border-start border-3 border-${getSeverityColor(alert.severity)} ps-3 mb-2">
                                <div class="d-flex justify-content-between">
                                    <strong>${alert.alert_type}</strong>
                                    <span class="badge bg-${getSeverityColor(alert.severity)}">${alert.severity}</span>
                                </div>
                                <div class="text-muted small">${alert.vehicle_name}</div>
                                <div class="text-muted small">${formatDateTime(alert.timestamp)}</div>
                            </div>
                        `).join('') : 
                        '<div class="text-muted text-center py-3">No recent alerts</div>'
                    }
                </div>
            </div>
        </div>
    `;
}

// Navigate to driver details page
function viewDriverPage(driverId = null) {
    const id = driverId || currentDriverId;
    if (id) {
        window.open(`/drivers/${id}`, '_blank');
    }
}

// Render pagination
function renderPagination(currentPage, totalPages, totalItems) {
    const paginationHtml = generatePaginationHtml(currentPage, totalPages, totalItems, 'loadPage');
    document.getElementById('driversPaginationTop').innerHTML = paginationHtml;
    document.getElementById('driversPaginationBottom').innerHTML = paginationHtml;
}

// Load specific page
function loadPage(page) {
    currentPage = page;
    loadDrivers();
}

// Update drivers count display
function updateDriversCount(total) {
    document.getElementById('driversCount').textContent = `${total} driver${total !== 1 ? 's' : ''} found`;
}

// Show/hide loading state
function showLoading(show) {
    const loading = document.getElementById('driversLoading');
    const table = document.getElementById('driversTable');
    
    if (show) {
        loading.style.display = 'block';
        table.style.opacity = '0.5';
    } else {
        loading.style.display = 'none';
        table.style.opacity = '1';
    }
}

// Refresh drivers
function refreshDrivers() {
    loadDrivers();
}

// Utility functions
function getAvatarColor(name) {
    const colors = [
        '#007bff', '#6f42c1', '#e83e8c', '#dc3545', '#fd7e14',
        '#ffc107', '#28a745', '#20c997', '#17a2b8', '#6c757d'
    ];
    const index = (name || '').charCodeAt(0) % colors.length;
    return colors[index];
}

function getSeverityColor(severity) {
    const colors = {
        'critical': 'danger',
        'high': 'warning',
        'medium': 'info',
        'low': 'success'
    };
    return colors[severity] || 'secondary';
}

function formatDate(dateString) {
    if (!dateString) return 'N/A';
    return new Date(dateString).toLocaleDateString();
}

function formatDateTime(dateString) {
    if (!dateString) return 'N/A';
    return new Date(dateString).toLocaleString();
}

function showError(message) {
    // You can implement a toast notification here
    console.error(message);
    alert('Error: ' + message);
}

// Generate pagination HTML (reusable function)
function generatePaginationHtml(currentPage, totalPages, totalItems, onClickFunction) {
    if (totalPages <= 1) return '';
    
    let html = '<nav><ul class="pagination pagination-sm justify-content-center mb-0">';
    
    // Previous button
    html += `<li class="page-item ${currentPage === 1 ? 'disabled' : ''}">
        <a class="page-link" href="#" onclick="${onClickFunction}(${currentPage - 1}); return false;">Previous</a>
    </li>`;
    
    // Page numbers
    const startPage = Math.max(1, currentPage - 2);
    const endPage = Math.min(totalPages, currentPage + 2);
    
    if (startPage > 1) {
        html += `<li class="page-item"><a class="page-link" href="#" onclick="${onClickFunction}(1); return false;">1</a></li>`;
        if (startPage > 2) {
            html += '<li class="page-item disabled"><span class="page-link">...</span></li>';
        }
    }
    
    for (let i = startPage; i <= endPage; i++) {
        html += `<li class="page-item ${i === currentPage ? 'active' : ''}">
            <a class="page-link" href="#" onclick="${onClickFunction}(${i}); return false;">${i}</a>
        </li>`;
    }
    
    if (endPage < totalPages) {
        if (endPage < totalPages - 1) {
            html += '<li class="page-item disabled"><span class="page-link">...</span></li>';
        }
        html += `<li class="page-item"><a class="page-link" href="#" onclick="${onClickFunction}(${totalPages}); return false;">${totalPages}</a></li>`;
    }
    
    // Next button
    html += `<li class="page-item ${currentPage === totalPages ? 'disabled' : ''}">
        <a class="page-link" href="#" onclick="${onClickFunction}(${currentPage + 1}); return false;">Next</a>
    </li>`;
    
    html += '</ul></nav>';
    
    // Add items info
    const startItem = (currentPage - 1) * 20 + 1;
    const endItem = Math.min(currentPage * 20, totalItems);
    html += `<div class="text-center text-muted mt-2 small">
        Showing ${startItem}-${endItem} of ${totalItems} drivers
    </div>`;
    
    return html;
} 